File: java-example/src/main/java/demo/WebSocketChatServer.java
package demo;

import java.util.Scanner;
import java.util.concurrent.CompletionStage;

import akka.actor.ActorSystem;
import akka.http.javadsl.Http;
import akka.http.javadsl.ServerBinding;
import akka.http.javadsl.model.ContentTypes;
import akka.http.javadsl.model.HttpEntities;
import akka.http.javadsl.model.StatusCodes;
import akka.http.javadsl.server.AllDirectives;
import akka.http.javadsl.server.Route;

/**
 * WebSocket Chat Server - Java Implementation
 * Demonstrates Java 21 with Akka HTTP server setup
 */
public class WebSocketChatServer extends AllDirectives {

    public static void main(String[] args) {
        ActorSystem system = ActorSystem.create("ChatWebSocketSystem");

        try {
            WebSocketChatServer server = new WebSocketChatServer();
            Route routes = server.createRoutes();

            CompletionStage<ServerBinding> bindingFuture = Http.get(system)
                .newServerAt("localhost", 8081)
                .bind(routes);

            bindingFuture.whenComplete((binding, failure) -> {
                if (binding != null) {
                    System.out.println("🚀 WebSocket Chat Server (Java) online at http://localhost:8081/");
                    System.out.println("🌐 Test page: http://localhost:8081/");
                    System.out.println("📊 Health check: http://localhost:8081/api/health");
                    System.out.println("Press RETURN to stop...");

                    Scanner scanner = new Scanner(System.in);
                    scanner.nextLine();

                    binding.unbind().whenComplete((unbound, unbindFailure) -> {
                        System.out.println("🛑 Server stopped");
                        system.terminate();
                    });

                } else {
                    System.err.println("❌ Failed to bind server: " + failure.getMessage());
                    system.terminate();
                }
            });

        } catch (Exception e) {
            System.err.println("❌ Server startup failed: " + e.getMessage());
            system.terminate();
        }
    }

    private Route createRoutes() {
        return route(
            pathPrefix("api", () -> route(
                path("health", () ->
                    get(() -> complete(StatusCodes.OK, HttpEntities.create(ContentTypes.APPLICATION_JSON, "{\"status\":\"healthy\"}"))
                    )
                ),
                path("info", () ->
                    get(() -> complete(StatusCodes.OK, HttpEntities.create(ContentTypes.APPLICATION_JSON,
                        "{\"service\":\"WebSocket Chat Server - Java\",\"version\":\"1.0.0\"}"))
                    )
                )
            )),
            pathSingleSlash(() ->
                complete(StatusCodes.OK, HttpEntities.create(ContentTypes.TEXT_HTML_UTF8, createTestPage()))
            )
        );
    }

    private String createTestPage() {
        return """
            <!DOCTYPE html>
            <html lang="en">
            <head>
                <meta charset="UTF-8">
                <title>Java WebSocket Chat Server</title>
                <style>
                    body { font-family: Arial, sans-serif; margin: 40px; background: #f5f5f5; }
                    .container { max-width: 800px; margin: 0 auto; background: white; padding: 30px; border-radius: 8px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); }
                    .status { background: #d4edda; color: #155724; padding: 15px; border-radius: 5px; margin: 20px 0; }
                    .java-badge { background: #f89406; color: white; padding: 4px 8px; border-radius: 4px; font-size: 12px; }
                    .endpoint { background: #e9ecef; padding: 10px; border-radius: 4px; margin: 10px 0; }
                    code { background: #f8f9fa; padding: 2px 6px; border-radius: 3px; }
                    pre { background: #f8f9fa; padding: 15px; border-radius: 5px; overflow-x: auto; }
                </style>
            </head>
            <body>
                <div class="container">
                    <h1>☕ Java WebSocket Chat Server <span class="java-badge">Java 21</span></h1>

                    <div class="status">
                        ✅ Java HTTP Server Running Successfully on Port 8081
                    </div>

                    <h2>🚀 Implementation Features</h2>
                    <ul>
                        <li><strong>Java 21</strong> with modern language features</li>
                        <li><strong>Akka HTTP 10.2.10</strong> server framework</li>
                        <li><strong>Actor Model</strong> foundation with Akka</li>
                        <li><strong>Maven</strong> build system</li>
                        <li><strong>Sealed Classes</strong> and pattern matching ready</li>
                    </ul>

                    <h2>🔗 Available Endpoints</h2>
                    <div class="endpoint">
                        <strong>Health Check:</strong>
                        <a href="/api/health" target="_blank">GET /api/health</a>
                    </div>
                    <div class="endpoint">
                        <strong>Server Info:</strong>
                        <a href="/api/info" target="_blank">GET /api/info</a>
                    </div>

                    <h2>🏃‍♂️ Running the Server</h2>
                    <pre>cd java-example
mvn clean compile exec:java -Dexec.mainClass="demo.WebSocketChatServer"</pre>

                    <h2>📝 Next Steps</h2>
                    <p>This server demonstrates the foundation for a Java WebSocket chat system. To add full WebSocket chat functionality:</p>
                    <ol>
                        <li>Implement WebSocket routing with <code>handleWebSocketMessages</code></li>
                        <li>Create chat actor system for message handling</li>
                        <li>Add JSON message processing with Jackson</li>
                        <li>Build real-time message broadcasting</li>
                    </ol>

                    <p><strong>Note:</strong> This Java server runs on port 8081 to avoid conflicts with the Scala implementation (port 8080).</p>

                    <h2>🎯 Language Comparison</h2>
                    <p>Compare this Java implementation with the Scala version running on port 8080 to see how both languages approach the Actor Model and WebSocket handling.</p>
                </div>
            </body>
            </html>
            """;
    }
}

File: scala-example/src/main/scala/demo/WebSocketChatServer.scala
package demo

import akka.actor.{ActorRef, ActorSystem, Props}
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import demo.chat.ChatSystem
import demo.http.routes.ChatRoutes

import scala.concurrent.duration._
import scala.concurrent.{ExecutionContext, Future}
import scala.io.StdIn
import scala.util.{Failure, Success}

/**
 * WebSocket Chat Server
 * HTTP server with WebSocket support for real-time multi-user chat testing
 */
object WebSocketChatServer {

  def main(args: Array[String]): Unit = {
    implicit val system: ActorSystem = ActorSystem("ChatWebSocketSystem")
    implicit val ec: ExecutionContext = system.dispatcher

    // Create chat system actor
    val chatSystem: ActorRef = system.actorOf(Props[ChatSystem](), "chat-system")

    // Create routes
    val chatRoutes = new ChatRoutes(chatSystem)

    val routes: Route = concat(
      chatRoutes.routes,
      // API routes for testing
      pathPrefix("api") {
        concat(
          path("health") {
            get {
              complete(HttpEntity(ContentTypes.`application/json`, """{"status":"healthy"}"""))
            }
          },
          path("info") {
            get {
              complete(HttpEntity(ContentTypes.`application/json`,
                """{"service":"WebSocket Chat Server","version":"1.0.0"}"""))
            }
          }
        )
      },
      // Serve simple HTML page for testing
      pathSingleSlash {
        complete(HttpEntity(ContentTypes.`text/html(UTF-8)`, createTestPage))
      }
    )

    val bindingFuture = Http().newServerAt("localhost", 8080).bind(routes)

    bindingFuture.onComplete {
      case Success(binding) =>
        println(s"🚀 WebSocket Chat Server online at http://localhost:8080/")
        println(s"📡 WebSocket endpoint: ws://localhost:8080/chat/ws/{username}")
        println(s"🌐 Test page: http://localhost:8080/")
        println(s"📊 Health check: http://localhost:8080/api/health")
        println(s"💬 Ready for multi-user chat testing!")
        println(s"Press RETURN to stop...")

        // Wait for user input to stop
        StdIn.readLine()

        binding
          .unbind()
          .onComplete { _ =>
            println("🛑 Server stopped")
            system.terminate()
          }

      case Failure(ex) =>
        println(s"❌ Failed to bind server: ${ex.getMessage}")
        system.terminate()
    }
  }

  private def createTestPage: String = {
    """<!DOCTYPE html>
      |<html lang="en">
      |<head>
      |    <meta charset="UTF-8">
      |    <meta name="viewport" content="width=device-width, initial-scale=1.0">
      |    <title>WebSocket Chat Test</title>
      |    <style>
      |        body { font-family: Arial, sans-serif; margin: 20px; background-color: #f5f5f5; }
      |        .container { max-width: 800px; margin: 0 auto; }
      |        .chat-container { background: white; border-radius: 8px; padding: 20px; margin: 20px 0; box-shadow: 0 2px 4px rgba(0,0,0,0.1); }
      |        .connection-status { padding: 10px; border-radius: 4px; margin: 10px 0; }
      |        .connected { background-color: #d4edda; color: #155724; }
      |        .disconnected { background-color: #f8d7da; color: #721c24; }
      |        .messages { height: 300px; overflow-y: auto; border: 1px solid #ddd; padding: 10px; margin: 10px 0; background: #f9f9f9; }
      |        .message { margin: 5px 0; padding: 8px; border-radius: 4px; }
      |        .message.sent { background: #e3f2fd; text-align: right; }
      |        .message.received { background: #f1f8e9; }
      |        .message.system { background: #fff3e0; font-style: italic; }
      |        input, button { padding: 8px; margin: 5px; border: 1px solid #ddd; border-radius: 4px; }
      |        button { background: #007bff; color: white; cursor: pointer; }
      |        button:hover { background: #0056b3; }
      |        button:disabled { background: #ccc; cursor: not-allowed; }
      |        .input-group { display: flex; margin: 10px 0; }
      |        .input-group input { flex: 1; }
      |    </style>
      |</head>
      |<body>
      |    <div class="container">
      |        <h1>🚀 WebSocket Chat Test</h1>
      |        <p>Test the chat system with multiple users by opening this page in multiple browser tabs/windows.</p>
      |
      |        <div class="chat-container">
      |            <div class="input-group">
      |                <input type="text" id="username" placeholder="Enter username" value="User1">
      |                <button id="connect">Connect</button>
      |                <button id="disconnect" disabled>Disconnect</button>
      |            </div>
      |
      |            <div id="status" class="connection-status disconnected">
      |                Status: Disconnected
      |            </div>
      |
      |            <div class="input-group">
      |                <input type="text" id="roomName" placeholder="Room name" value="general">
      |                <button id="joinRoom" disabled>Join Room</button>
      |                <button id="listRooms" disabled>List Rooms</button>
      |                <button id="listUsers" disabled>List Users</button>
      |            </div>
      |
      |            <div id="messages" class="messages"></div>
      |
      |            <div class="input-group">
      |                <input type="text" id="messageInput" placeholder="Type your message..." disabled>
      |                <button id="sendMessage" disabled>Send</button>
      |            </div>
      |        </div>
      |
      |        <div class="chat-container">
      |            <h3>📋 Instructions for Multi-User Testing</h3>
      |            <ol>
      |                <li>Open this page in multiple browser tabs/windows</li>
      |                <li>Set different usernames in each tab (e.g., Alice, Bob, Charlie)</li>
      |                <li>Click "Connect" in each tab</li>
      |                <li>Join the same room (e.g., "general") in all tabs</li>
      |                <li>Start sending messages from different tabs</li>
      |                <li>Observe real-time message delivery across all connected users</li>
      |            </ol>
      |        </div>
      |    </div>
      |
      |    <script>
      |        let socket = null;
      |        let username = '';
      |        let currentRoom = '';
      |
      |        const elements = {
      |            username: document.getElementById('username'),
      |            connect: document.getElementById('connect'),
      |            disconnect: document.getElementById('disconnect'),
      |            status: document.getElementById('status'),
      |            roomName: document.getElementById('roomName'),
      |            joinRoom: document.getElementById('joinRoom'),
      |            listRooms: document.getElementById('listRooms'),
      |            listUsers: document.getElementById('listUsers'),
      |            messages: document.getElementById('messages'),
      |            messageInput: document.getElementById('messageInput'),
      |            sendMessage: document.getElementById('sendMessage')
      |        };
      |
      |        function addMessage(message, type = 'system') {
      |            const messageEl = document.createElement('div');
      |            messageEl.className = `message ${type}`;
      |            messageEl.textContent = message;
      |            elements.messages.appendChild(messageEl);
      |            elements.messages.scrollTop = elements.messages.scrollHeight;
      |        }
      |
      |        function updateStatus(connected) {
      |            elements.status.textContent = connected ? `Status: Connected as ${username}` : 'Status: Disconnected';
      |            elements.status.className = `connection-status ${connected ? 'connected' : 'disconnected'}`;
      |
      |            elements.connect.disabled = connected;
      |            elements.disconnect.disabled = !connected;
      |            elements.joinRoom.disabled = !connected;
      |            elements.listRooms.disabled = !connected;
      |            elements.listUsers.disabled = !connected;
      |            elements.messageInput.disabled = !connected;
      |            elements.sendMessage.disabled = !connected;
      |        }
      |
      |        function connect() {
      |            username = elements.username.value.trim();
      |            if (!username) {
      |                alert('Please enter a username');
      |                return;
      |            }
      |
      |            socket = new WebSocket(`ws://localhost:8080/chat/ws/${encodeURIComponent(username)}`);
      |
      |            socket.onopen = function() {
      |                updateStatus(true);
      |                addMessage(`Connected as ${username}`, 'system');
      |            };
      |
      |            socket.onmessage = function(event) {
      |                try {
      |                    const data = JSON.parse(event.data);
      |                    handleMessage(data);
      |                } catch (e) {
      |                    addMessage(`Raw message: ${event.data}`, 'system');
      |                }
      |            };
      |
      |            socket.onclose = function() {
      |                updateStatus(false);
      |                addMessage('Disconnected', 'system');
      |            };
      |
      |            socket.onerror = function(error) {
      |                addMessage(`Error: ${error}`, 'system');
      |            };
      |        }
      |
      |        function disconnect() {
      |            if (socket) {
      |                socket.close();
      |                socket = null;
      |            }
      |        }
      |
      |        function handleMessage(data) {
      |            switch (data.type) {
      |                case 'joinedRoom':
      |                    currentRoom = data.roomName;
      |                    addMessage(`Joined room: ${data.roomName}`, 'system');
      |                    break;
      |                case 'messageSent':
      |                    addMessage(`You: ${data.content}`, 'sent');
      |                    break;
      |                case 'messageReceived':
      |                    addMessage(`${data.username}: ${data.content}`, 'received');
      |                    break;
      |                case 'roomList':
      |                    addMessage(`Rooms: ${data.rooms.join(', ')}`, 'system');
      |                    break;
      |                case 'userList':
      |                    addMessage(`Users: ${data.users.join(', ')}`, 'system');
      |                    break;
      |                case 'error':
      |                    addMessage(`Error: ${data.error}`, 'system');
      |                    break;
      |                case 'heartbeat':
      |                    // Ignore heartbeat messages
      |                    break;
      |                default:
      |                    addMessage(`Unknown message: ${JSON.stringify(data)}`, 'system');
      |            }
      |        }
      |
      |        function joinRoom() {
      |            const roomName = elements.roomName.value.trim();
      |            if (!roomName || !socket) return;
      |
      |            socket.send(JSON.stringify({
      |                type: 'joinRoom',
      |                roomName: roomName
      |            }));
      |        }
      |
      |        function sendMessage() {
      |            const message = elements.messageInput.value.trim();
      |            if (!message || !socket || !currentRoom) return;
      |
      |            socket.send(JSON.stringify({
      |                type: 'sendMessage',
      |                roomName: currentRoom,
      |                content: message
      |            }));
      |
      |            elements.messageInput.value = '';
      |        }
      |
      |        function listRooms() {
      |            if (!socket) return;
      |            socket.send(JSON.stringify({type: 'listRooms'}));
      |        }
      |
      |        function listUsers() {
      |            if (!socket) return;
      |            socket.send(JSON.stringify({type: 'listUsers'}));
      |        }
      |
      |        // Event listeners
      |        elements.connect.addEventListener('click', connect);
      |        elements.disconnect.addEventListener('click', disconnect);
      |        elements.joinRoom.addEventListener('click', joinRoom);
      |        elements.listRooms.addEventListener('click', listRooms);
      |        elements.listUsers.addEventListener('click', listUsers);
      |        elements.sendMessage.addEventListener('click', sendMessage);
      |
      |        elements.messageInput.addEventListener('keypress', function(e) {
      |            if (e.key === 'Enter') {
      |                sendMessage();
      |            }
      |        });
      |
      |        elements.username.addEventListener('keypress', function(e) {
      |            if (e.key === 'Enter') {
      |                connect();
      |            }
      |        });
      |
      |        elements.roomName.addEventListener('keypress', function(e) {
      |            if (e.key === 'Enter') {
      |                joinRoom();
      |            }
      |        });
      |    </script>
      |</body>
      |</html>""".stripMargin
  }
}
